src/state/token_index.mli

```diff
@@ -0,0 +1,11 @@
+val sync_dir : string ref
+val mutex_json_token : Lwt_mutex.t
+val cast_list : 'a -> [> `List of 'a ]
+val string_to_json : string -> Yojson.Basic.t
+val token_to_string : Learnocaml_data.Token.t list -> string list
+val get_file : string -> unit -> Yojson.Basic.t Lwt.t
+val get_tokens : unit -> Learnocaml_data.Token.t list Lwt.t
+val list_cast : string list -> Yojson.Basic.t list
+val write_index :  string list -> unit Lwt.t
+val create_index : unit Lwt.t
+val add_token : Learnocaml_data.Token.t -> unit -> unit Lwt.t
```

Comment:
you have been pretty exhaustive by naming here the functions provided in `token_index.ml`

however, it would be a good idea to remove most of these: the main advantage of an interface being that all functions that are not listed here, will be hidden from the module user (e.g. `sync_dir` and `mutex_json_token` are certainly intended to be "private", so they should be hidden this way, and most functions that are just helper functions, can be hidden just as well).

Finally, you may want to add some documentation in this `.mli`:
To sum up, I suggest just putting in this file at most:

```suggestion
(* This file is part of Learn-OCaml.
 *
 * Copyright (C) 2019-2020 OCaml Software Foundation.
 * Copyright (C) 2016-2018 OCamlPro.
 *
 * Learn-OCaml is distributed under the terms of the MIT license. See the
 * included LICENSE file for details. *)

(** Create or regenerate token index from sync/ and write sync/token.json.
    This step may take a long time (up to several minutes).  Automatically
    called (once and for all) by [get_tokens] or [add_token] if need be. *)
val create_index : unit Lwt.t

(** Get the list of all tokens. *)
val get_tokens : unit -> Learnocaml_data.Token.t list Lwt.t

(** Add a registered token in the index. *)
val add_token : Learnocaml_data.Token.t -> unit -> unit Lwt.t
```

Reply:
The sync_dir function have his utility : the token's module need it to write/read data, but the learnocaml_store.ml need it too. 
Instead of changing twice the sync_dir if you have some modification (And possibly forget to change one), learnocaml_store.ml pick his sync_dir directly by this function.